spopt/locate/p_median.py

```diff
@@ -533,3 +535,576 @@ def solve(self, solver: pulp.LpSolver, results: bool = True):
+        >>> k_nearest_pmedian = k_nearest_pmedian.solve(pulp.PULP_CBC_CMD(msg=False)
+
+        Get the facility-client associations.
+
+        >>> for fac, cli in enumerate(k_nearest_pmedian.fac2cli):
+        ...     print(f"facility {fac} serving {len(cli)} clients")
+        facility 0 serving 1 clients
+        facility 1 serving 1 clients
+        facility 2 serving 0 clients
+
+        Get the total and average weighted travel cost.
+
+        >>> round(k_nearest_pmedian.problem.objective.value(), 3)
+        1.618
+        >>> round(k_nearest_pmedian.mean_dist, 3)
+        0.809
+
+        Get the k list for the last iteration.
+        >>> print(prob.k_list)
+        [2, 1]
+
+        """
+
+        # check the crs of two geodataframes
+        if gdf_demand.crs is None:
+            raise ValueError("GeoDataFrame gdf_demand does not have a valid CRS.")
+        if gdf_fac.crs is None:
+            raise ValueError("GeoDataFrame gdf_facility does not have a valid CRS.")
+        if gdf_demand.crs != gdf_fac.crs:
+            raise ValueError(
+                "Geodataframes crs are different: "
+                f"gdf_demand-{gdf_demand.crs}, gdf_fac-{gdf_fac.crs}"
+            )
+
+        # create the array of coordinate of clients and facilities
+        dem = gdf_demand[demand_col]
+        fac = gdf_fac[facility_col]
+        dem_data = np.array([dem.x.to_numpy(), dem.y.to_numpy()]).T
+        fac_data = np.array([fac.x.to_numpy(), fac.y.to_numpy()]).T
+
+        # demand and capacity
+        service_load = gdf_demand[weights_cols].to_numpy()
+        facility_capacities = None
+        if facility_capacity_col is not None:
+            facility_capacities = gdf_fac[facility_capacity_col].to_numpy()
+
+        cls.clients = dem_data
+        cls.facilities = fac_data
+        cls.weights = service_load
+        cls.capacities = facility_capacities
+        cls.k_list = k_list
+        cls.distance_metric = distance_metric
+        cls.p_facilities = p_facilities
+        cls.name = name
+
+        return cls
+
+    def facility_client_array(self) -> None:
+        """
+
+        Create a 2D array storing **facility to client relationships** where each
+        row represents a facility and contains an array of client indices with
+        which it is associated. An empty client array indicates the facility
+        is associated with no clients.
+
+        Returns
+        -------
+
+        None
+        """
+        fac_vars = getattr(self, "fac_vars")
+        cli_vars = getattr(self, "cli_assgn_vars")
+        len_fac_vars = len(fac_vars)
+
+        self.fac2cli = []
+
+        for j in range(len_fac_vars):
+            array_cli = []
+            if fac_vars[j].value() > 0:
+                for i in range(len(cli_vars)):
+                    if (i, j) in cli_vars and cli_vars[i, j].value() > 0:
+                        array_cli.append(i)
+
+            self.fac2cli.append(array_cli)
+
+    @classmethod
```

Comment:
this `@classmethod` needs to be removed. `solve()` should be an instance method, since we use it from a "initialised" object.